official/resnet/resnet_run_loop.py

```diff
@@ -563,24 +584,41 @@ def input_fn_eval():
     schedule = [flags_obj.epochs_between_evals for _ in range(int(n_loops))]
     schedule[-1] = flags_obj.train_epochs - sum(schedule[:-1])  # over counting.
 
+  use_train_and_evaluate = isinstance(
+      distribution_strategy, tf.contrib.distribute.CollectiveAllReduceStrategy)
   for cycle_index, num_train_epochs in enumerate(schedule):
     tf.compat.v1.logging.info('Starting cycle: %d/%d', cycle_index,
                               int(n_loops))
 
     if num_train_epochs:
-      classifier.train(input_fn=lambda: input_fn_train(num_train_epochs),
-                       hooks=train_hooks, max_steps=flags_obj.max_train_steps)
-
-    tf.compat.v1.logging.info('Starting to evaluate.')
+      if use_train_and_evaluate:
+        train_spec = tf.estimator.TrainSpec(
+            input_fn=lambda: input_fn_train(num_train_epochs), hooks=train_hooks,
+            max_steps=flags_obj.max_train_steps)
+      else:
+        classifier.train(input_fn=lambda: input_fn_train(num_train_epochs),
+                         hooks=train_hooks, max_steps=flags_obj.max_train_steps)
+    elif use_train_and_evaluate:
+      train_spec = tf.estimator.TrainSpec(
+          input_fn=lambda: None, hooks=train_hooks,
+          max_steps=flags_obj.max_train_steps)
 
     # flags_obj.max_train_steps is generally associated with testing and
     # profiling. As a result it is frequently called with synthetic data, which
     # will iterate forever. Passing steps=flags_obj.max_train_steps allows the
     # eval (which is generally unimportant in those circumstances) to terminate.
     # Note that eval will run for max_train_steps each loop, regardless of the
     # global_step count.
-    eval_results = classifier.evaluate(input_fn=input_fn_eval,
-                                       steps=flags_obj.max_train_steps)
+    if use_train_and_evaluate:
+      eval_spec = tf.estimator.EvalSpec(input_fn=input_fn_eval,
+                                        steps=flags_obj.max_train_steps)
+      tf.compat.v1.logging.info('Starting to train and evaluate.')
+      eval_results, _ = tf.estimator.train_and_evaluate(classifier, train_spec,
```

Comment:
I am not sure whether it is the best practice to call `train_and_evaluate` in a loop. Usually we make an input_fn for all epochs and training finishes when `train_and_evaluate` returns. Also `train_and_evaluate` would run evaluation on evaluator continuously until it reaches  `max_train_steps`.

Reply:
Can you elaborate a bit more? Does this mean when using `train_and_evaluate` there is no loop? But then how do we specify number of epochs?

Reply:
We specify the number of epochs in the `input_fn` or specify the number of steps in `train_spec`. @karmel can correct me if I understand it wrong.

Reply:
My understanding matches @yuefengz 's -- train and eval runs the loop for you.

Reply:
Thanks, done.